File: exercise/ts/day13/src/santamarket.model/ReceiptItem.ts
import { Product } from './Product';

export class ReceiptItem {
    constructor(
        public product: Product,
        public quantity: number,
        public price: number,
        public totalPrice: number
    ) {}

    equals(other: ReceiptItem): boolean {
        return (
            other instanceof ReceiptItem &&
            this.product.equals(other.product) &&
            Math.abs(this.price - other.price) < 0.001 &&
            Math.abs(this.totalPrice - other.totalPrice) < 0.001 &&
            Math.abs(this.quantity - other.quantity) < 0.001
        );
    }
}

File: exercise/kotlin/day13/src/main/kotlin/santamarket/model/ReceiptItem.kt
package santamarket.model

import kotlin.math.abs

data class ReceiptItem(
    val product: Product,
    val quantity: Double,
    val price: Double,
    val totalPrice: Double
) {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ReceiptItem) return false

        val tolerance = 0.001
        return product == other.product &&
                abs(price - other.price) < tolerance &&
                abs(totalPrice - other.totalPrice) < tolerance &&
                abs(quantity - other.quantity) < tolerance
    }

    override fun hashCode(): Int = arrayOf(product, price, totalPrice, quantity).contentHashCode()
}